Tests/CacheAdvanceTests/CacheAdvanceTests.swift

```diff
@@ -96,31 +96,108 @@ final class CacheAdvanceTests: XCTestCase {
+            return CacheAdvance<TestableMessage>(
+                fileURL: testFileLocation,
+                writer: try FileHandle(forWritingTo: testFileLocation),
+                reader: try CacheReader(
+                    forReadingFrom: testFileLocation,
+                    maximumBytes: maximumBytes),
+                header: header,
+                decoder: JSONDecoder(),
+                encoder: JSONEncoder())
+        }
+        let writingCache = try makeCache()
+        try writingCache.append(message: message)
+
+        // Make the file corrupted by setting the offset at end of newest message to be further in the file.
+        // This could happen if a crash occurred during a write of `header.offsetInFileAtEndOfNewestMessage` on a big-endian device.
+        try header.updateOffsetInFileAtEndOfNewestMessage(
+            to: header.offsetInFileAtEndOfNewestMessage + 1)
+
+        // Create a new cache instance that uses the corrupted data persisted to disk
+        let corruptedReadingCache = try makeCache()
+
+        XCTAssertThrowsError(try corruptedReadingCache.messages()) {
+            XCTAssertEqual($0 as? CacheAdvanceError, CacheAdvanceError.fileCorrupted)
+        }
+    }
+
+    func test_messages_throwsFileCorruptedWhenOffsetInFileAtEndOfNewestMessageIsBeforeEndOfNewestMessage() throws {
+        let message: TestableMessage = "This is a test"
+        let maximumBytes = try requiredByteCount(for: [message])
+        let header = try CacheHeaderHandle(
+            forReadingFrom: testFileLocation,
+            maximumBytes: maximumBytes,
+            overwritesOldMessages: true)
+
+        func makeCache() throws -> CacheAdvance<TestableMessage> {
+            return CacheAdvance<TestableMessage>(
+                fileURL: testFileLocation,
+                writer: try FileHandle(forWritingTo: testFileLocation),
+                reader: try CacheReader(
+                    forReadingFrom: testFileLocation,
+                    maximumBytes: maximumBytes),
+                header: header,
+                decoder: JSONDecoder(),
+                encoder: JSONEncoder())
+        }
+        let writingCache = try makeCache()
+        try writingCache.append(message: message)
+
+        // Make the file corrupted by setting the offset at end of newest message to be earlier in the file.
+        // This could happen if a crash occurred during a write of `header.offsetInFileAtEndOfNewestMessage` on a little-endian device.
         try header.updateOffsetInFileAtEndOfNewestMessage(
-            to: FileHeader.expectedEndOfHeaderInFile + 1)
+            to: header.offsetInFileAtEndOfNewestMessage - 1)
+
+        // Create a new cache instance that uses the corrupted data persisted to disk
+        let corruptedReadingCache = try makeCache()
 
-        XCTAssertThrowsError(try cache.messages()) {
+        XCTAssertThrowsError(try corruptedReadingCache.messages()) {
```

Comment:
To make sure I follow again... what was the behavior here before we updated the library?

Reply:
I should have left a comment! The behavior of this test before the fix was an infinite loop – this test would not complete.